### PR TITLE
Add MFA grant type

### PR DESCRIPTION
**Pull request by DInuwan97**
**Completed**

- Change the project structure.
- Use authentication context instead of direct user objects in Event Listeners.
- Enable file config capability to configure one custom authenticator.(This should have to improve based on removing hard-coded Basic authenticator and Identifier first step)
- Changes have done as per the recently finalized API Spec.
- Custom extensions are added to customize the MFA authentication based on user logged in attribute.
- Change the DAO layer and enable cache layers and implement a data persistence layer.
- Configurable error messages and error codes in property files.
- Table schema and cleanup stored procedures available for both MSSQL and MySQL databases.
- Can implement any event handler by using WSO2 Eventing framework and from the event handlers, this connector is having the ability to pass error message to the API response.

**Todo**

- Keep the same flowId, until the flow is completed

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth2.grant.rest.framework/src/main/java/org/wso2/carbon/identity/oauth2/grant/rest/framework/cache/AuthCache.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+
+package org.wso2.carbon.identity.oauth2.grant.rest.framework.cache;
+import org.wso2.carbon.identity.core.cache.BaseCache;
+public class AuthCache extends BaseCache<AuthCacheKey, AuthCacheEntry> {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth2.grant.rest.framework/src/main/java/org/wso2/carbon/identity/oauth2/grant/rest/framework/cache/AuthCache.java:22]: Formatting

File: README.md [hunk 2]
```diff
@@ -1,2 +1,201 @@
...
+Supported Versions : 5.10.0, 5.11.0, 6.0.0, 6.1.0
+
+This connector has developed to provide an RESTful API interface for 
+multi-factor authentication with WSO2 Identity Server. The connector itself 
+contains its own grant type to get an access token at the end of multi-factor 
+authentication steps. This would support Identifier First, Basic 
+Authenticator, SMS OTP and Email OTP authenticators with any number 
+of authentication steps.
+
+As per the [API spec](https://app.swaggerhub.com/apis/KALUBOWILADC25132/wso-2_identity_server_api_authentication_connector/v1.0)  this connector contains 3 endpoints namely
```
Comment (janakamarasena) [inline README.md:12]: Point to the yaml in the repo.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,201 @@
-# identity-oauth2-grant-mfa
-identity-oauth2-grant-mfa is a REST API that enforces an OAuth2 grant type according to a set of multi factor authentication steps.
+# WSO2 Identity Server - API Authentication Connector
+
+Supported Versions : 5.10.0, 5.11.0, 6.0.0, 6.1.0
+
+This connector has developed to provide an RESTful API interface for
```
Comment (janakamarasena) [inline README.md:5]: lets add a diagram showing the overall flow and a brief intro.

Comment (janakamarasena) [reply README.md:5]: Ideally reading the readme anyone should understand completely what this connector is, how to deploy it and how to use it.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,201 @@
...
+default_access = "allow"
+context = "(.*)/api/identity/authn/v1/auth-steps"
+secure = "false"
+http_method = "GET"
+
+[tenant_context.rewrite]
+custom_webapps=["/api/identity/authn/v1/"]
+```
+
+#### SMSOTP and EmailOTP Configurations
```
Comment (janakamarasena) [inline README.md:167]: are these configs specific for this connector? if not add a link to the official docs of the configs

File: README.md [hunk 5]
```diff
@@ -1,2 +1,201 @@
...
+
+4) Copy `identity.oauth2.grant.auth.rest.handler-x.x.x.jar` file into the
+`<IS-Home>/repository/components/libs` directory.
+
+5) Get the `org.wso2.carbon.extension.identity.smsotp.common-x.x.x.jar` and 
+`org.wso2.carbon.extension.identity.emailotp.common-x.x.x.jar` and copy in to 
+the `<IS-Home>/repository/components/dropins` directory. 
+(The 02 JARs are required to this connector in order to send the OTP for 
+respective channel if [EmailOTP](https://github.com/wso2-extensions/identity-outbound-auth-sms-otp/tree/master/component/common) 
+or [SMSOTP](https://github.com/wso2-extensions/identity-outbound-auth-email-otp/tree/master/component/common)
```
Comment (rksk) [inline README.md:127]:
```suggestion
respective channel if [SMSOTP](https://github.com/wso2-extensions/identity-outbound-auth-sms-otp/tree/master/component/common) 
or [EmailOTP](https://github.com/wso2-extensions/identity-outbound-auth-email-otp/tree/master/component/common) 
```

File: components/org.wso2.carbon.identity.oauth2.grant.rest.endpoint/pom.xml [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+<!--                    </execution>-->
+<!--                </executions>-->
+<!--                <dependencies>-->
+<!--                    <dependency>-->
+<!--                        <groupId>org.openapitools</groupId>-->
+<!--                        <artifactId>cxf-wso2-openapi-generator</artifactId>-->
+<!--                        <version>1.0.0</version>-->
+<!--                    </dependency>-->
+<!--                </dependencies>-->
+<!--            </plugin>-->
```
[the remark this answers is not in the document]
Comment (DInuwan97) [reply components/org.wso2.carbon.identity.oauth2.grant.rest.endpoint/pom.xml]: We can remove this, but this is the dependency which generate the gen code of the endpoint. Hence do we need to remove or comment out?

File: docs/README.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+3) Copy `org.wso2.carbon.identity.oauth2.grant.rest.core-x.x.x.jar`,
+   `org.wso2.carbon.extension.identity.smsotp.common-x.x.x.jar` and
+   `org.wso2.carbon.extension.identity.emailotp.common-x.x.x.jar` files into the 
+`<IS-Home>/repository/components/dropins` directory.
+
+4) Copy `identity.oauth2.grant.auth.rest.handler-x.x.x.jar` file into the
+`<IS-Home>/repository/components/libs` directory.
+
+5) Copy `api#identity#authn#v1.war` file into the `<IS-Home>/repository/deployment/server` directory.
```
Comment (pabasara-mahindapala) [inline docs/README.md:131]:
```suggestion
5) Copy `api#identity#authn#v1.war` file into the `<IS-Home>/repository/deployment/server/webapps` directory.
```

**Review by pabasara-mahindapala — CHANGES_REQUESTED** [review]
Correct the webapps directory

File: components/org.wso2.carbon.identity.oauth2.grant.rest.endpoint/src/main/java/org/wso2/carbon/identity/oauth2/grant/rest/endpoint/util/ConfigUtil.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
+package org.wso2.carbon.identity.oauth2.grant.rest.endpoint.util;
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.oauth2.grant.rest.endpoint/src/main/java/org/wso2/carbon/identity/oauth2/grant/rest/endpoint/util/ConfigUtil.java:1]: There are multiple files with no license header. Can you check all and fix them?

Comment (DInuwan97) [reply components/org.wso2.carbon.identity.oauth2.grant.rest.endpoint/src/main/java/org/wso2/carbon/identity/oauth2/grant/rest/endpoint/util/ConfigUtil.java:19]: fixed in 733ee6bbfe71c91b9d1c6ad17be4cd4feb38fcbc

**Review by ashensw — APPROVED** [review]

**Comment by ashensw** [comment]
Note: Approved the PR due to the urgency, based on the code reviews conducted to review the main flows of the implementation and relevant codes associated with them. Unit tests need to be added as further improvements.

**Pull request merged**